hr_timesheet_activity_begin_end/models/account_analytic_line.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+#
+#
+#    Authors: Guewen Baconnier
+#    Copyright 2015 Camptocamp SA
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+from __future__ import division
+
+import math
+from datetime import timedelta
+
+from openerp import models, fields, api, exceptions, _
+from openerp.tools.float_utils import float_compare
+
+
+def float_time_convert(float_val):
+    hours = math.floor(abs(float_val))
+    mins = abs(float_val) - hours
+    mins = round(mins * 60)
+    if mins >= 60.0:
+        hours = hours + 1
+        mins = 0.0
+    return '%02d:%02d' % (hours, mins)
+
+
+class AccountAnalyticLine(models.Model):
+    _inherit = 'account.analytic.line'
+
+    time_start = fields.Float(string='Begin Hour')
+    time_stop = fields.Float(string='End Hour')
+
+    @api.one
+    @api.constrains('time_start', 'time_stop', 'unit_amount')
+    def _check_time_start_stop(self):
+        start = timedelta(hours=self.time_start)
+        stop = timedelta(hours=self.time_stop)
+        if stop < start:
+            raise exceptions.ValidationError(
+                _('The beginning hour (%s) must '
+                  'precede the ending hour (%s).') %
+                (float_time_convert(self.time_start),
+                 float_time_convert(self.time_stop))
+            )
+        hours = (stop - start).seconds / 3600
+        if (hours and
+                float_compare(hours, self.unit_amount, precision_digits=4)):
+            raise exceptions.ValidationError(
+                _('The duration (%s) must be equal to the difference '
+                  'between the hours (%s).') %
+                (float_time_convert(self.unit_amount),
+                 float_time_convert(hours))
+            )
+        # check if lines overlap
+        others = self.search([
+            ('id', '!=', self.id),
+            ('user_id', '=', self.user_id.id),
+            ('date', '=', self.date),
+            '|',
+            '|',
+            '&',
+            ('time_start', '<', self.time_start),
+            ('time_stop', '>', self.time_start),
+            '&',
+            ('time_start', '<', self.time_stop),
+            ('time_stop', '>', self.time_stop),
+            '&',
+            ('time_start', '>', self.time_start),
+            ('time_stop', '<', self.time_stop),
```

Comment:
Theorem: given 2 segments S1 = [A1, B1] and S2 = [A2, B2], S1 and S2 overlap if and only if A1 < B2 && A2 < B1

Demonstration

case 1:

    A1             B1
    |-------------|
           |--------------|
           A2              B2

Case 2:

    A1                             B1
    |---------------------------|
           |--------------|
           A2              B2

And the expression is symetric which is nice because if S1 overlaps S2 then obviously S2 overlaps S1. 


Could you please simplify the domain with this in mind? I believe the resulting expression will be much simpler.